File: LeetCode/303.Range_Sum_Query_-_Immutable/Mario/303.Range_Sum_Query_-_Immutable.java
class NumArray {

    private int[] sumArr;

    public NumArray(int[] nums) {
        sumArr = new int[nums.length];
        for (int i = 0; i < nums.length; i++) {
            sumArr[i] = nums[i];
            if (i - 1 >= 0) {
                sumArr[i] += sumArr[i - 1];
            }
        }
        // System.out.println(Arrays.toString(sumArr));
    }

    public int sumRange(int left, int right) {
        int sum = sumArr[right];
        if (left > 0) {
            sum -= sumArr[left - 1];
        }
        return sum;
    }
}

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray obj = new NumArray(nums);
 * int param_1 = obj.sumRange(left,right);
 */

File: LeetCode/303.Range_Sum_Query_-_Immutable/303.Range_Sum_Query_-_Immutable.swift

class NumArray {
    var prefixSum = [0]

    init(_ nums: [Int]) {
        for num in nums {
            if let last = prefixSum.last {
                prefixSum.append(num + last)
            }
        }
    }

    func sumRange(_ left: Int, _ right: Int) -> Int {
        return prefixSum[right + 1] - prefixSum[left]
    }
}

/**
 * Your NumArray object will be instantiated and called as such:
 * let obj = NumArray(nums)
 * let ret_1: Int = obj.sumRange(left, right)
 */
